File: dsa-with-subeen-sir/algorithom/binary_search/binary_search.c
#include <stdio.h>

// Function to perform Binary Search
int binary_search(int arr[], int size, int target) {
    int left, right, mid;
    left = 0;   // First index of the array
    right = size - 1;  // Last index of the array

    while(left <= right) { // Continue while search space exists
        // Calculate middle index
        // Safer than (left + right) / 2 for large values
        mid = left + (right - left) / 2;

        if(arr[mid] == target) {  // If target is found
            return mid;
        } else if(target < arr[mid]) { //  Target is smaller than middle element so Search only in the left half
            right = mid - 1;  // right decrease
        } else { // Target is greater than middle element so  Search only in the right half
            left = mid + 1; // left increase
        }
    }

    return -1;   // Target not found
}

int main() {
    // Sorted array (Required)
    int arr[] = {5, 10, 15, 20, 25, 30, 35};

    // Calculate total number of elements
    int size = sizeof(arr) / sizeof(arr[0]);

    int target = 25;
    
    // Call Binary Search function
    int result = binary_search(arr, size, target);

    // Print result
    if (result != -1)
        printf("Element found at index %d\n", result);
    else
        printf("Element not found\n");

    return 0;
}

File: dsa-with-subeen-sir/algorithom/binary_search/binary_search.py
def binary_search(arr, target):
    # Start searching from the first index
    left = 0

    # End searching at the last index
    right = len(arr) - 1

    # Continue searching while the search space exists
    while left <= right:
        # Find the middle index
        # is integer division (returns whole number)
        mid = (left + right) // 2

        if arr[mid] == target:  # If middle element is the target, return position
            return mid

        # যদি মাঝের element (arr[mid]) target-এর থেকে ছোট হয়,
        # তাহলে target শুধুমাত্র ডান পাশেই থাকতে পারে (কারণ array sorted)।
        # তাই বাম পাশের সব element (mid সহ) বাদ দিয়ে,
        # left-কে mid-এর পরের index (mid + 1) থেকে শুরু করি।
        if arr[mid] < target:
            left = mid + 1

        # অন্যথায় arr[mid] অবশ্যই target-এর থেকে বড়।
        # তাই target শুধুমাত্র বাম পাশেই থাকতে পারে।
        # ডান পাশের সব element (mid সহ) বাদ দিয়ে,
        # right-কে mid-এর আগের index (mid - 1) এ নিয়ে আসি।
        else:
            right = mid - 1

    # If loop finishes, target doesn't exist
    return -1


# Sorted array (Required)
numbers = [5, 10, 15, 20, 25, 30, 35]

target = 25  
# target = 5

result = binary_search(numbers, target)

if result != -1:
    print("Element found at index:", result)
else:
    print("Element not found")
